atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -2509,38 +2564,81 @@ private void logCommitLockTenureExceeded(
             }
+            rowLockCount += lockDescriptors.size() - previousLockCount;
         }
-        result.add(AtlasRowLockDescriptor.of(
+        lockDescriptors.add(AtlasRowLockDescriptor.of(
                 TransactionConstants.TRANSACTION_TABLE.getQualifiedName(),
                 TransactionConstants.getValueForTimestamp(getStartTimestamp())));
         cellCommitLocksRequested = cellLockCount;
         rowCommitLocksRequested = rowLockCount + 1;
-        return result;
+        return LocksAndMetadata.of(
+                lockDescriptors,
+                // For now, lock request metadata only consists of change metadata. If it is absent, we can save
+                // computation by not doing an index encoding
+                lockDescriptorToChangeMetadata.isEmpty()
+                        ? Optional.empty()
+                        : Optional.of(LockRequestMetadata.of(lockDescriptorToChangeMetadata)));
+    }
+
+    private void collectCellLocks(
+            Set<LockDescriptor> lockDescriptors,
+            Map<LockDescriptor, ChangeMetadata> lockDescriptorToChangeMetadata,
+            TableReference tableRef) {
+        Map<Cell, ChangeMetadata> changeMetadataForWrites = getChangeMetadataForWrites(tableRef);
+        for (Cell cell : getLocalWrites(tableRef).keySet()) {
+            LockDescriptor lockDescriptor =
+                    AtlasCellLockDescriptor.of(tableRef.getQualifiedName(), cell.getRowName(), cell.getColumnName());
+            lockDescriptors.add(lockDescriptor);
+            if (changeMetadataForWrites.containsKey(cell)) {
+                lockDescriptorToChangeMetadata.put(lockDescriptor, changeMetadataForWrites.get(cell));
+            }
+        }
+    }
+
+    private void collectRowLocks(
+            Set<LockDescriptor> lockDescriptors,
+            Map<LockDescriptor, ChangeMetadata> lockDescriptorToChangeMetadata,
+            TableReference tableRef) {
+        Map<Cell, ChangeMetadata> changeMetadataForWrites = getChangeMetadataForWrites(tableRef);
+        Cell lastCell = null;
+        for (Cell cell : getLocalWrites(tableRef).keySet()) {
+            LockDescriptor rowLockDescriptor =
+                    AtlasRowLockDescriptor.of(tableRef.getQualifiedName(), cell.getRowName());
+            if (changeMetadataForWrites.containsKey(cell)) {
+                if (lockDescriptorToChangeMetadata.containsKey(rowLockDescriptor)) {
+                    throw new SafeIllegalStateException(
+                            "Two different cells in the same row have metadata and we create locks on row level.",
+                            UnsafeArg.of("tableRef", tableRef),
+                            UnsafeArg.of("rowName", cell.getRowName()),
+                            UnsafeArg.of("existingMetadata", lockDescriptorToChangeMetadata.get(rowLockDescriptor)),
+                            UnsafeArg.of("newMetadata", changeMetadataForWrites.get(cell)));
```

Comment:
This is another spicy bit that I did no get to in the RFC. Because we are using `SERIALIZABLE` conflict handler, we must transfer metadata from cells to rows. A direct consequence of this is that we cannot allow multiple metadata objects per row. 

This is a bit of a hidden invariant. The alternative is to throw in `putWithMetadata` already (this does not get rid of the above snipped, but would rather be an additional check). However, this requires maintaining some additional data structure to track metadata on a row basis. Also, guaranteeing atomicity in this context is probably harder (and uglier as the checks for duplicate metadata would also have to be performed atomically with everything else). 

If we are fine with noting down this invariant in the documentation for `putWithMetadata` and having `commit` fail instead, our life is a bit easier.

Reply:
I think it's fine for us to throw on commit, given that this is before we actually acquire locks _AND_ write to the KVS